src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -554,6 +554,62 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
   CallGenerator::do_late_inline_helper();
 }
 
+static bool has_non_debug_usages(Node* n) {
+  for (DUIterator_Fast imax, i = n->fast_outs(imax); i < imax; i++) {
+    Node* m = n->fast_out(i);
+    if (!m->is_SafePoint()
+        || (m->is_Call() && m->as_Call()->has_non_debug_use(n))) {
+      return true;
+    }
+  }
+  return false;
+}
+
+static void scalarize_debug_usages(CallNode* call, Node* resproj) {
+  Unique_Node_List safepoints;
+  for (DUIterator_Fast imax, i = resproj->fast_outs(imax); i < imax; i++) {
+    Node* m = resproj->fast_out(i);
+    safepoints.push(m);
+  }
+
+#ifndef PRODUCT
+  if (PrintEliminateAllocations && safepoints.size() > 0) {
+    tty->print("++++ Eliminated: %d ", call->_idx);
+    call->as_CallStaticJava()->method()->print_short_name(tty);
+    tty->cr();
+  }
+#endif
+
+  GraphKit kit(call->jvms());
+  PhaseGVN& gvn = kit.gvn();
+  // delay box in runtime, treat box as a scalarized object
+  while (safepoints.size() > 0) {
+    ProjNode* res = resproj->as_Proj();
+    Node* sfpt = safepoints.pop();
+
+    ciInstanceKlass* klass = call->as_CallStaticJava()->method()->holder();
+    int n_fields = klass->nof_nonstatic_fields();
+    assert(n_fields == 1, "the klass must be an auto-boxing klass");
+
+    uint first_ind = sfpt->req() - sfpt->jvms()->scloff();
+    Node* sobj = new SafePointScalarObjectNode(gvn.type(res)->isa_oopptr(),
+#ifdef ASSERT
+                                                  call->isa_Allocate(),
```

Comment:
It is always `NULL` since `call` can't be an `Allocate`.

Reply:
It was my suggestion to pass `call` node as allocation so that we could trace back for what node SafePointScalarObject was created because you may have several Box objects for which we create SafePointScalarObject nodes.
I think we can change argument (and field type) to CallNode. And have assert in SafePointScalarObject constructor to check that it is either Allocate node or Boxing Call node.

Reply:
Thank you for your review. I will change the  SafePointScalarObject constructor in next push.